File: dotnet/src/Core/AggregateRoot.cs
namespace Core;

public abstract class AggregateRoot
{
    public string AggregateId { get; }

    private readonly List<IEvent<object>> _uncommitted = [];

    protected AggregateRoot(string aggregateId)
    {
        AggregateId = aggregateId;
    }

    protected abstract void Apply(IEvent<object> @event);

    protected void Raise(IEvent<object> @event)
    {
        Apply(@event);
        _uncommitted.Add(@event);
    }

    public IReadOnlyList<IEvent<object>> GetUncommittedEvents() => _uncommitted.AsReadOnly();

    public void Replay(IEnumerable<IEvent<object>> history)
    {
        foreach (var e in history)
            Apply(e);
    }
}

File: node/src/core/AggregateRoot.ts
import { Event } from './Event'

export abstract class AggregateRoot {
  private readonly _uncommittedEvents: Event[] = []
  protected _aggregateId!: string

  constructor(aggregateId: string) {
    this._aggregateId = aggregateId
  }

  get aggregateId(): string {
    return this._aggregateId
  }

  get uncommittedEvents() : Event[] {
    return this._uncommittedEvents
  }

  protected raise(event: Event): void {
    this.apply(event)
    this._uncommittedEvents.push(event)
  }

  replay(history: Event[]): void {
    history.forEach(e => this.apply(e))
  }

  abstract apply(event: Event): void
}